pkg/controller/etcdproxy/certs.go

```diff
@@ -7,198 +7,350 @@ import (
+		}
+	}
+
+	return utilerrors.NewAggregate(errs)
+}
+
+// ensureServerCertificates handles certificate generating, renewal and rotation for Serving CA bundle and Server certificates.
+// The Serving CA bundle is saved in a ConfigMaps defined in EtcdStorage Spec.
+// The Server certificate/key pair is stored in the Secrets named etcdstorageName-server-cert in the controller namespace.
+// The EtcdProxy controller assumes ConfigMaps for the Serving CA bundle are already created, but if not, the controller will try to create them.
+// Creating ConfigMaps for storing the Serving CA bundle requires the appropriate RBAC roles if RBAC is enabled on cluster.
+//
+// This function reads the Secret in the controller namespace and checks is the 'etcd.xmudrii.com/certificate-expiry-date'
+// annotation present and contains valid date. If date is in the past, i.e. certificate is expired, or the annotation is not present controller:
+// * Generates new CA certificate. The new CA certificate is appended to all ConfigMaps specified by the EtcdStorage Spec.
+// Expired CA certificates from the bundle are removed in this phase.
+// * Generates new Server certificate/key pair using the newly generated CA certificate and update Secret in the controller namespace with new pair.
+// * "Restarts" etcd-proxy to pick-up new changes.
+// TODO: Implement "restarting" etcd-proxy. This is going to be done using rolling updates.
+func (c *EtcdProxyController) ensureServerCertificates(etcdstorage *etcdstoragev1alpha1.EtcdStorage) error {
+	serverSecret, err := c.getSeverCertSecret(etcdstorage)
 	if err != nil {
 		return err
 	}
-	serverSignerCert, _, err := serverSigner.GetPEMBytes()
+
+	// Check is annotation containing expiry date present and valid. If it is valid, we're skipping this iteration.
+	if expiry, ok := serverSecret.Annotations[ProxyCertificateExpiryAnnotation]; ok {
+		certExpiry, err := time.Parse(time.RFC3339, expiry)
+		if err != nil {
+			return err
+		}
+		if certExpiry.After(time.Now()) {
+			return nil
+		}
+	}
+
+	// Generate new Serving CA certificate and Server certificate/key pair.
+	servingCA, serverCert, err := c.generateServerBundle(etcdstorage)
 	if err != nil {
 		return err
 	}
 
-	// Generate server certificate/key pair.
-	serverBundle, err := serverSigner.NewServerCertificate(pkix.Name{CommonName: serviceUrl},
-		[]string{serviceUrl}, r.Int63n(100000), currentTime)
+	// Write new Server certificate/key pair to the Secret in the controller namespace.
+	serverCertBytes, serverKeyBytes, err := serverCert.GetPEMBytes()
 	if err != nil {
 		return err
 	}
-	serverCertBytes, serverKeyBytes, err := serverBundle.GetPEMBytes()
+
+	serverSecret.Annotations = map[string]string{
+		ProxyCertificateExpiryAnnotation: serverCert.Certificates[0].NotAfter.Format(time.RFC3339),
+	}
+	serverSecret.Data = map[string][]byte{
+		"tls.crt": serverCertBytes,
+		"tls.key": serverKeyBytes,
+	}
+	err = ensureSecret(c.kubeclientset, serverSecret)
```

Comment:
Doesn't this make it possible to start serving with a serving cert that none of your clients honor?

Reply:
Looks like yes. What would be a good way to fix this? Move it down after updating client ConfigMaps? But in that case could there be similar problem?